### PR TITLE
refactor the forward sampling for srp_model 

### PR DESCRIPTION
**Pull request by AKktk**
The main objective is re-implementation in c++ to avoid delay of looping in R

### CODE REVIEW

File: src/impute_srp_model.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            from_out.push_back("stable");
+            to_out.push_back("response");
+            t_min_out.push_back(tmin);
+            t_max_out.push_back(tmax);
+            t_sot_out.push_back(t_sot(i));
+            iter_out.push_back(iter);
+            //append to results vectors response -> progression
+            subject_id_out.push_back(subject_id(i));
+            group_id_out.push_back(group_id(i));
+            from_out.push_back("stable");
```
Comment (kkmann) [inline src/impute_srp_model.cpp]: shouldnt this be response?

Comment (AKktk) [reply src/impute_srp_model.cpp]: Yep, corrected it

File: src/impute_srp_model.cpp [hunk 2]
```diff
@@ -148,9 +148,29 @@ DataFrame impute_srp_model(
...
         }
         if (from(i) == "response") {
-
-          // todo
-
+          if (from[i - 1] != "stable" ||
+              subject_id[i - 1] != subject_id[i]) {
+            stop("Unexpected");
+          }
+          double t_response = rtruncweibull(
```
Comment (AKktk) [inline src/impute_srp_model.cpp:166]: @kkmann, Why do we use the t_min and t_max values from i-1 here?

File: src/impute_srp_model.cpp [hunk 3]
```diff
@@ -148,9 +148,29 @@ DataFrame impute_srp_model(
...
+            shape[iter][g][0], scale[iter][g][0], t_min(i-1), t_max(i-1)
+          );
+          double dt_progression = rtruncweibull(
+              shape[iter][g][2], scale[iter][g][2],
+                                               t_min(i) - t_response, R_PosInf
+          );
+          double n_visits_prog = floor(dt_progression / visit_spacing(g));
```
Comment (AKktk) [inline src/impute_srp_model.cpp:173]: @kkmann, 
I believe our code will be much faster, if we don't do variable declaration within the loop. Instead do it once, outside

File: src/impute_srp_model.cpp (hunk 2 above)
Comment (kkmann) [reply src/impute_srp_model.cpp:166]: if you are coming "from response" there needs to be a stable-> response in the line before. You then sample response jump first (i-1) and then the progression jump (i). I now realize that this could be dangerous if the previous line is not correctly aligned in the data frame. Should probably add a check for that (e.g. that the subject ids are the same and the previous line is indeed a "from stable")

File: src/impute_srp_model.cpp (hunk 3 above)
Comment (kkmann) [reply src/impute_srp_model.cpp:173]: good point +1. Also, we might need to pre-allocate the _out vectors to avoid dynamic allocation of memory. Let's do that in a next PR though so that we don't pack too much in this one.

File: R/srp_model.R [hunk 4]
```diff
@@ -134,159 +134,21 @@ is_valid.srp_model <- function(mdl) { # nolint
...
-      )
-  }
-  if (debug == TRUE) {
-    return(temp)
-  } else {
-    return(res)
-  }
+  # TODO: can we fix that by "correctly" indexing in the matrix
+  p <- matrix(p[idx, ], nrow = 1)
+  temp <- impute_srp_model(
```
Comment (kkmann) [inline R/srp_model.R]: temp -> res is convention

File: R/srp_model.R [hunk 5]
```diff
@@ -134,159 +134,21 @@ is_valid.srp_model <- function(mdl) { # nolint
...
+    data,
+    p,
+    as.vector(shape[idx, , ]),
+    as.vector(scale[idx, , ]),
+    visit_spacing,
+    nsim,
+    as.integer(length(group_id_levels))
+  ) %>%
+    arrange(subject_id, iter) %>%
```
Comment (kkmann) [inline R/srp_model.R]: indentation

**Comment by kkmann** [comment]
next steps: correct issues with short response->progression time (same t_min/t_max for s->r and r->p) and completing the test set.

File: tests/testthat/test-srp_model_priorpredictive.R [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      p_se = sd(responder) / sqrt(n())
+    )
+
+  means <- tbl_cpp %>%
+    filter(from == "stable", to == "response") %>%
+    summarise(
+      mean_hat = mean((t_min + t_max) / 2),
+      mean_se = sd((t_min + t_max) / 2) / sqrt(n())
+    )
+  expect_true(all((means$mean_hat - 4.33 * gamma(2)) < 0.1))
```
Comment (kkmann) [inline tests/testthat/test-srp_model_priorpredictive.R]: @AKktk could you add the rationale for 4.33 and some comments to explain what you are doing/testing here please?

File: tests/testthat/test_impute.R [hunk 7]
```diff
@@ -0,0 +1,67 @@
+test_that("Testing .impute function", {
```
Comment (kkmann) [inline tests/testthat/test_impute.R]: can you be a bit more specific? it is ok to have more than on test_that() per file

File: tests/testthat/test_impute.R [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    median_time_to_next_event = matrix(c(
+      3, 3, 6,
+      3, 3, 6,
+      3, 3, 6
+    ), byrow = TRUE,  nrow = 3, ncol = 3),
+    median_time_to_next_event_sd = matrix(1, byrow = TRUE,  nrow = 3, ncol = 3)
+  )
+  smpl_prior <- sample_prior(mdl, warmup = 500, nsim = 2000, seed = 36L)
+
+  # sample only from one group
```
Comment (kkmann) [inline tests/testthat/test_impute.R]: that's not what is happening, you sample from two groups!

File: tests/testthat/test_impute.R [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                            seed = 3423423)
+  tbl2 <- sample_predictive(mdl,
+                            sample = smpl_prior,
+                            n_per_group = c(1L, 1L, 0L),
+                            nsim = 20,
+                            seed = 3423423)
+  expect_true(all(tbl1 == tbl2))
+  # TODO: Take care of the corner case where progression and response intervals
+  # are the same
+  # TODO: Write a test case with low median time, so that many corner cases are
```
Comment (kkmann) [inline tests/testthat/test_impute.R]: I think you can remove the todo, please split this into its's own test_that() and add comments explaining what you are doing/testing

File: tests/testthat/test_impute.R [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                            sample = smpl_prior2,
+                            n_per_group = c(5L),
+                            nsim = 100,
+                            seed = 3423423)
+  corner_check <- tbl3 %>%
+         group_by(iter, subject_id) %>%
+         filter(t_min == lag(t_min) & t_max == lag(t_max)) %>%
+    nrow()
+  expect_true(corner_check == 0)
+    # TODO: check that only one individual from group one is sampled once
```
Comment (kkmann) [inline tests/testthat/test_impute.R]: this is not addressed yet if I see it correctly

**Review by kkmann — CHANGES_REQUESTED** [review]
Yeah, CI says it works - almost there!

File: tests/testthat/test_impute.R (hunk 10 above)
Comment (AKktk) [reply tests/testthat/test_impute.R]: 'Only one individual is sampled' part?
What is to be checked there can you please elaborate a bit more on that?

File: tests/testthat/test_impute.R (hunk 9 above)
Comment (AKktk) [reply tests/testthat/test_impute.R]: Alright

File: tests/testthat/test-srp_model_priorpredictive.R (hunk 6 above)
Comment (kkmann) [reply tests/testthat/test-srp_model_priorpredictive.R]: done myself

File: tests/testthat/test_srp_model_impute.R [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      mean_se = sd((t_min + t_max) / 2) / sqrt(n())
+    )
+  # work out the theoretical mean given scale = 1 and specified median = 3
+  # (see mdl definition and https://en.wikipedia.org/wiki/Weibull_distribution)
+  theoretical_mean <- mdl$median_time_to_next_event_mean[1]/log(2) * gamma(2)
+  # testing for comparison with theoretical mean, allowing for estimation error
+  expect_true(with(tbl_observed_mean,
+    abs(mean_hat - theoretical_mean) <= 2 * mean_se
+  ))
+  # TODO: Check other two means, can you do that elegantly without copy-paste?
```
Comment (kkmann) [inline tests/testthat/test_srp_model_impute.R]: @AKktk can u look into this?

File: src/impute_srp_model.cpp [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+          // first sample whether response or progression happens
+          double pr_response = conditional_response_probability_srp(
+              t_min(i) - t_sot(i), // (min) time since start of treatment
+              p(iter, g), // unconditional response probability
+              // transition index -1 due to 0 based indexing in C
+              shape[iter][g][0], shape[iter][g][1],
+              scale[iter][g][0], scale[iter][g][1]
+            );
+          bool response = R::rbinom(1, pr_response);
+          if (response) {
```
Comment (kkmann) [inline src/impute_srp_model.cpp:95]: this seems to be where the current discrepancy comes from @AKktk almost certain it is an offset issue

Comment (AKktk) [reply src/impute_srp_model.cpp:95]: checked the if(response){ } block

**Review by kkmann — APPROVED** [review]
👍🏻
